apps/passport-server/src/services/poapService.ts

```diff
@@ -0,0 +1,253 @@
+import { getEdDSAPublicKey } from "@pcd/eddsa-pcd";
+import { getHash } from "@pcd/passport-crypto";
+import { SerializedPCD } from "@pcd/pcd-types";
+import { ZKEdDSAEventTicketPCDPackage } from "@pcd/zk-eddsa-event-ticket-pcd";
+import AsyncLock from "async-lock";
+import { fetchDevconnectPretixTicketByTicketId } from "../database/queries/devconnect_pretix_tickets/fetchDevconnectPretixTicket";
+import {
+  claimNewPoapUrl,
+  getExistingClaimUrlByTicketId
+} from "../database/queries/poap";
+import { ApplicationContext } from "../types";
+import { logger } from "../util/logger";
+import { getServerErrorUrl } from "../util/util";
+import { RollbarService } from "./rollbarService";
+import { traced } from "./telemetryService";
+
+// Set up an async-lock to prevent concurrency issues when two threads
+// are simultaneously calling `getDevconnectPoapClaimUrl()`.
```

Comment:
To clarify - nodejs is single-threaded. The concurrency issue is slightly different.